src/cue.ml

```diff
@@ -0,0 +1,244 @@
+  mutable value : 'a;
+  mutable capacity : int;
+  mutable counter : int;
+}
+
+external next_as_atomic : 'a node -> 'a node Atomic.t = "%identity"
+
+let[@inline] get_next node = Atomic.get (next_as_atomic node)
+let[@inline] fenceless_get_next node = fenceless_get (next_as_atomic node)
+
+external as_record : 'a node -> 'a record = "%identity"
+
+let[@inline] get_capacity node = (as_record node).capacity
+let[@inline] set_capacity node value = (as_record node).capacity <- value
+let[@inline] get_counter node = (as_record node).counter
+let[@inline] set_counter node value = (as_record node).counter <- value
+
+let[@inline] compare_and_set_next node before after =
+  Atomic.compare_and_set (next_as_atomic node) before after
+
+type 'a t = {
+  head : 'a node Atomic.t;
+  head_waiters : (unit -> unit) list Atomic.t;
+  capacity : int;
+  tail_waiters : (unit -> unit) list Atomic.t;
+  tail : 'a node Atomic.t;
+}
+
+let create ?(capacity = Int.max_int) () =
+  let value = Obj.magic () in
+  let node = Node { _next = Null; value; capacity; counter = 0 } in
+  let head = Atomic.make node
+  and head_waiters = Atomic.make []
+  and tail_waiters = Atomic.make []
+  and tail = Atomic.make node in
+  { head; head_waiters; capacity; tail_waiters; tail }
+
+let capacity_of t = t.capacity
+
+let is_empty t =
+  let head = Atomic.get t.head in
+  fenceless_get_next head == Null
+
+let rec snapshot t =
+  let head = Atomic.get t.head in
+  let tail = fenceless_get t.tail in
+  match fenceless_get_next tail with
+  | Node _ as node ->
+      Atomic.compare_and_set t.tail tail node |> ignore;
+      snapshot t
+  | Null -> if Atomic.get t.head != head then snapshot t else (head, tail)
+
+let length t =
+  let head, tail = snapshot t in
+  get_counter tail - get_counter head
+
+(* *)
+
+let rec release_all waiters =
+  let releases = fenceless_get waiters in
+  if releases != [] then
+    if Atomic.compare_and_set waiters releases [] then
+      List.iter (fun release -> release ()) releases
+    else release_all waiters
+
+(* *)
+
+let rec peek t =
+  let old_head = Atomic.get t.head in
+  match fenceless_get_next old_head with
+  | Null ->
+      let dla = Domain_local_await.prepare_for_await () in
+      let releases = Atomic.get t.tail_waiters in
+      if Atomic.compare_and_set t.tail_waiters releases (dla.release :: releases)
+      then (
+        if old_head != Atomic.get t.tail then release_all t.tail_waiters
+        else
+          try dla.await ()
+          with exn ->
+            release_all t.tail_waiters;
+            raise exn);
+      peek t
+  | Node r as node ->
+      let value = r.value in
+      if Atomic.get t.head != node then peek t else value
+
+let[@inline] peek t = peek t
+
+(* *)
+
+let rec peek_opt t =
+  let head = Atomic.get t.head in
+  match fenceless_get_next head with
+  | Null -> None
+  | Node r as node ->
+      let value = r.value in
+      if Atomic.get t.head != node then peek_opt t else Some value
```

Comment:
` if Atomic.get t.head != node then ...` 

There is a small bug here, it should be : 

 `if Atomic.get t.head != head then ...`

Same in `peek`.

Reply:
There is still some bugs between `peek_opt` and `pop_opt`.

                                  |                        
                        Try_push 42 : true               
                                  |                        
               .------------------------------------.
               |                                    |                        
      Peek_opt : Some (0)            Pop_opt : Some (42) 

I will continue investigating !

Reply:
Good catch!  Yes, this is a bug.  What would ideally be needed here is a(n acquire) fence to prevent the `r.value` read from being reordered after the `Atomic.get`, but OCaml currently doesn't provide a direct way to do _only_ that.   Some alternatives:
- `value` could be turned into an atomic and then `r.value` would be replaced with `Atomic.get r.value`.   This would be quite expensive.
- Instead of `Atomic.get t.head == head` one could use `Atomic.compare_and_set t.head head head`.  This is also somewhat expensive, because means that the `t.head` location is actually written to.
- In practise, but not in the OCaml memory model, which doesn't match actual hardware, it is also possible to generate fences by accessing an unrelated location.  For example, one could define
  ```ocaml
  let fence () = Atomic.set (Atomic.make 0) 0
  ```
  and use that to write
  ```ocaml
  let value = r.value in
  fence ();
  if Atomic.get t.head != head then ...
  ```

Reply:
The `fence` trick seems to work !

Reply:
This may actually be a compiler bug.

The OCaml 5 compiler [definitely has a bug](https://github.com/ocaml/ocaml/issues/12713) related to an incorrect optimization of `Atomic.get` operations.

What remains to consider is whether it is actually allowed to reorder a nonatomic load that precedes an atomic load to happen after the atomic load.  I suspect it is not actually allowed.  See the explanation of fences in this [post](https://preshing.com/20130922/acquire-and-release-fences/) about fences in C++11:

> An acquire fence prevents the memory reordering of any read which precedes it in program order with any read or write which follows it in program order.

This, of course, is not exactly the same as the semantics of `Atomic.get` in OCaml 5, but according to the above explanation the original code

```ocaml
let value = r.value in
(* The above read should not be allowed to happen after
    the read below as the read below has an acquire fence. *)
if Atomic.get t.head != head then ...
```

would be correct.  I'll try to redecipher the [Bounding Data Races in Space and Time](https://kcsrk.info/papers/pldi18-memory.pdf) paper to find definite word on this.

Reply:
Assuming this is just a compiler bug, then the following workaround would be sufficient:

```ocaml
let value = r.value in
(* The [Sys.opaque_identity] below prevents OCaml 5 from
    optimizing the repeated load away. *)
if Atomic.get (Sys.opaque_identity t.head) != head then ...
```

Reply:
Just to summarize, before I mark this as resolved, this turned out to be a compiler bug, which has been fixed in OCaml 5.2.